File: analytics.py
import sqlite3
import pandas as pd

DATABASE_NAME = "inventory.db"


def get_analytics():

    connection = sqlite3.connect(DATABASE_NAME)

    # Read products table into a Pandas DataFrame
    df = pd.read_sql_query("SELECT * FROM products", connection)

    connection.close()

    # If database has no products
    if df.empty:
        return {
            "total_products": 0,
            "total_inventory_value": 0,
            "low_stock_products": [],
            "products_by_category": {},
            "most_valuable_products": [],
            "restock_recommendations": []
        }

    # -----------------------------------
    # Total Products
    # -----------------------------------
    total_products = len(df)

    # -----------------------------------
    # Total Inventory Value
    # quantity × purchase_price
    # -----------------------------------
    df["inventory_value"] = df["quantity"] * df["purchase_price"]

    total_inventory_value = float(df["inventory_value"].sum())

    # -----------------------------------
    # Low Stock Products
    # -----------------------------------
    low_stock = df[
        df["quantity"] <= df["reorder_level"]
    ][
        ["product_name", "quantity", "reorder_level"]
    ]

    low_stock_products = low_stock.to_dict(orient="records")

    # -----------------------------------
    # Products By Category
    # -----------------------------------
    products_by_category = (
        df.groupby("category")
          .size()
          .to_dict()
    )

    # -----------------------------------
    # Most Valuable Products
    # -----------------------------------
    valuable = (
        df.sort_values(
            by="inventory_value",
            ascending=False
        )
        .head(5)
    )

    most_valuable_products = valuable[
        ["product_name", "inventory_value"]
    ].to_dict(orient="records")

    # -----------------------------------
    # Restock Recommendations
    # -----------------------------------
    restock = df[
        df["quantity"] <= df["reorder_level"]
    ][
        ["product_name", "quantity", "reorder_level"]
    ]

    restock_recommendations = restock.to_dict(orient="records")

    return {

        "total_products": total_products,

        "total_inventory_value": total_inventory_value,

        "low_stock_products": low_stock_products,

        "products_by_category": products_by_category,

        "most_valuable_products": most_valuable_products,

        "restock_recommendations": restock_recommendations
    }

File: frontend/src/pages/Analytics.js
import { useEffect, useState } from "react";
import { getAnalytics } from "../services/Api";

function Analytics() {

    const [analytics, setAnalytics] = useState(null);

    useEffect(() => {

        loadAnalytics();

    }, []);

    const loadAnalytics = () => {

        getAnalytics()
            .then((response) => {

                setAnalytics(response.data);

            })
            .catch((error) => {

                console.log(error);

            });

    };

    if (!analytics) {

        return (
            <div className="p-8">
                Loading Analytics...
            </div>
        );

    }

    return (

    <div className="p-8 bg-gray-100 min-h-screen">

        <h1 className="text-4xl font-bold mb-8">
            📊 Inventory Analytics
        </h1>

        {/* Summary Cards */}

        <div className="grid grid-cols-3 gap-6 mb-8">

            <div className="bg-blue-600 text-white rounded-lg shadow-lg p-6">

                <h2 className="text-lg">Total Products</h2>

                <p className="text-4xl font-bold mt-3">
                    {analytics.total_products}
                </p>

            </div>

            <div className="bg-green-600 text-white rounded-lg shadow-lg p-6">

                <h2 className="text-lg">Inventory Value</h2>

                <p className="text-3xl font-bold mt-3">
                    ₹ {analytics.total_inventory_value.toLocaleString()}
                </p>

            </div>

            <div className="bg-red-600 text-white rounded-lg shadow-lg p-6">

                <h2 className="text-lg">Low Stock Products</h2>

                <p className="text-4xl font-bold mt-3">
                    {analytics.low_stock_products.length}
                </p>

            </div>

        </div>

        {/* Products by Category */}

        <div className="bg-white rounded-lg shadow-lg p-6 mb-8">

            <h2 className="text-2xl font-semibold mb-4">
                Products by Category
            </h2>

            <table className="w-full">

                <thead className="bg-gray-200">

                    <tr>

                        <th className="p-3 text-left">Category</th>

                        <th className="p-3 text-left">Products</th>

                    </tr>

                </thead>

                <tbody>

                    {Object.entries(analytics.products_by_category).map(([category, count]) => (

                        <tr key={category} className="border-b">

                            <td className="p-3">{category}</td>

                            <td className="p-3">{count}</td>

                        </tr>

                    ))}

                </tbody>

            </table>

        </div>

        {/* Most Valuable Products */}

        <div className="bg-white rounded-lg shadow-lg p-6 mb-8">

            <h2 className="text-2xl font-semibold mb-4">
                Top Valuable Products
            </h2>

            <table className="w-full">

                <thead className="bg-gray-200">

                    <tr>

                        <th className="p-3 text-left">Product</th>

                        <th className="p-3 text-left">Inventory Value</th>

                    </tr>

                </thead>

                <tbody>

                    {analytics.most_valuable_products.map((product, index) => (

                        <tr key={index} className="border-b">

                            <td className="p-3">
                                {product.product_name}
                            </td>

                            <td className="p-3">
                                ₹ {product.inventory_value.toLocaleString()}
                            </td>

                        </tr>

                    ))}

                </tbody>

            </table>

        </div>

        {/* Restock Recommendations */}

        <div className="bg-white rounded-lg shadow-lg p-6">

            <h2 className="text-2xl font-semibold mb-4">
                Restock Recommendations
            </h2>

            {analytics.restock_recommendations.length === 0 ? (

                <p className="text-green-600 font-semibold">
                    🎉 No products need restocking.
                </p>

            ) : (

                <ul className="list-disc pl-6">

                    {analytics.restock_recommendations.map((product, index) => (

                        <li key={index}>
                            {product.product_name}
                        </li>

                    ))}

                </ul>

            )}

        </div>

    </div>

    );

}

export default Analytics;